File: 参考/产物/Balatro_1.0.1o/源码/engine/particles.lua
---@class Particles: Moveable
Particles = Moveable:extend()

--Class Methods
function Particles:init(X, Y, W, H, config)
    config = config or {}

    Moveable.init(self,X, Y, W, H)

    self.fill = config.fill
    self.padding = config.padding or 0

    if config.attach then
        self:set_alignment{
            major = config.attach,
            type = 'cm',
            bond = 'Strong'
        }
        table.insert(self.role.major.children,self)
        self.parent = self.role.major
        self.T.x = self.role.major.T.x + self.padding
        self.T.y = self.role.major.T.y + self.padding
        if self.fill then
            self.T.w = self.role.major.T.w - self.padding
            self.T.h = self.role.major.T.h - self.padding
        end
    end

    self.states.hover.can = false
    self.states.click.can = false
    self.states.collide.can = false
    self.states.drag.can = false
    self.states.release_on.can = false

    self.timer = config.timer or 0.5
    self.timer_type = (self.created_on_pause and 'REAL') or config.timer_type or 'REAL'
    self.last_real_time = G.TIMERS[self.timer_type] - self.timer
    self.last_drawn = 0
    self.lifespan = config.lifespan or 1
    self.fade_alpha = 0
    self.speed = config.speed or 1
    self.max = config.max or 1000000000000000
    self.pulse_max = math.min(20, config.pulse_max or 0)
    self.pulsed = 0
    self.vel_variation = config.vel_variation or 1
    self.particles = {}
    self.scale = config.scale or 1
    self.colours = config.colours or {G.C.BACKGROUND.D}

    if config.initialize then
        for i = 1, 60 do
            self.last_real_time = self.last_real_time - 15/60
            self:update(15/60)
            self:move(15/60)
        end
    end

    if getmetatable(self) == Particles then 
        table.insert(G.I.MOVEABLE, self)
    end
end

function Particles:update(dt)
    if G.SETTINGS.paused and not self.created_on_pause then self.last_real_time = G.TIMERS[self.timer_type] ; return end
    local added_this_frame = 0
    while G.TIMERS[self.timer_type]  > self.last_real_time + self.timer and (#self.particles < self.max or self.pulsed < self.pulse_max) and added_this_frame < 20 do
        self.last_real_time = self.last_real_time + self.timer
        local new_offset = { 
            x=self.fill and (0.5-math.random())*self.T.w or 0,
            y=self.fill and (0.5-math.random())*self.T.h or 0
        }
        if self.fill and self.T.r < 0.1 and self.T.r > -0.1 then 
            local newer_offset = {
                x = math.sin(self.T.r)*new_offset.y + math.cos(self.T.r)*new_offset.x,
                y = math.sin(self.T.r)*new_offset.x + math.cos(self.T.r)*new_offset.y,
            }
            new_offset = newer_offset
        end
        table.insert(self.particles, {
            draw = false,
            dir = math.random()*2*math.pi,
            facing = math.random()*2*math.pi,
            size = math.random()*0.5+0.1,
            age = 0,
            velocity = self.speed*(self.vel_variation*math.random() + (1-self.vel_variation))*0.7,
            r_vel = 0.2*(0.5 - math.random()),
            e_prev = 0,
            e_curr = 0,
            scale = 0,
            visible_scale = 0,
            time = G.TIMERS[self.timer_type],
            colour = pseudorandom_element(self.colours),
            offset = new_offset
        })
        added_this_frame = added_this_frame + 1
        if self.pulsed <= self.pulse_max then self.pulsed = self.pulsed + 1 end
    end
end

function Particles:move(dt)
    if G.SETTINGS.paused and not self.created_on_pause then return end

    Moveable.move(self, dt)

    if self.timer_type ~= 'REAL' then dt = dt*G.SPEEDFACTOR end

    for i=#self.particles,1,-1 do
        self.particles[i].draw = true
        self.particles[i].e_vel = self.particles[i].e_vel or dt*self.scale
        self.particles[i].e_prev = self.particles[i].e_curr
        self.particles[i].age = self.particles[i].age + dt
        
        self.particles[i].e_curr = math.min(2*math.min((self.particles[i].age/self.lifespan)*self.scale, self.scale*((self.lifespan - self.particles[i].age)/self.lifespan)), self.scale)

        self.particles[i].e_vel =  (self.particles[i].e_curr - self.particles[i].e_prev)*self.scale*dt + (1-self.scale*dt)*self.particles[i].e_vel

        self.particles[i].scale = self.particles[i].scale + self.particles[i].e_vel
        self.particles[i].scale = math.min(2*math.min((self.particles[i].age/self.lifespan)*self.scale, self.scale*((self.lifespan - self.particles[i].age)/self.lifespan)), self.scale)

        if self.particles[i].scale < 0 then
            table.remove(self.particles, i)
        else
            self.particles[i].offset.x = self.particles[i].offset.x + self.particles[i].velocity*math.sin(self.particles[i].dir)*dt
            self.particles[i].offset.y = self.particles[i].offset.y + self.particles[i].velocity*math.cos(self.particles[i].dir)*dt
            self.particles[i].facing = self.particles[i].facing + self.particles[i].r_vel*dt
            self.particles[i].velocity = math.max(0, self.particles[i].velocity - self.particles[i].velocity*0.07*dt)
        end
    end
end

function Particles:fade(delay, to)
    G.E_MANAGER:add_event(Event({
        trigger = 'ease',
        timer = self.timer_type, 
        blockable = false, 
        blocking = false,
        ref_value = 'fade_alpha',
        ref_table = self,
        ease_to = to or 1,
        delay = delay
    }))
end

function Particles:draw(alpha)
    alpha = alpha or 1
        prep_draw(self, 1)
        love.graphics.translate(self.T.w/2, self.T.h/2)
        for k, v in pairs(self.particles) do
            if v.draw then 
                love.graphics.push()
                love.graphics.setColor(v.colour[1], v.colour[2], v.colour[3], v.colour[4]*alpha*(1-self.fade_alpha))                
                love.graphics.translate(v.offset.x, v.offset.y)
                love.graphics.rotate(v.facing)
                
                love.graphics.rectangle('fill', -v.scale/2, -v.scale/2, v.scale, v.scale) -- origin in the middle
                love.graphics.pop()
            end
        end
        love.graphics.pop()

        add_to_drawhash(self)
        self:draw_boundingrect()
end

function Particles:remove()
    if self.role.major then 
        for k, v in pairs(self.role.major.children) do
            if v == self and type(k) == 'number' then
                table.remove(self.role.major.children, k)
            end
        end
    end

    remove_all(self.children)

    Moveable.remove(self)
end

File: 复刻/Balatro/src/game/particles.ts
/**
 * `engine/particles.lua` 的直译（22 号票）：开包时的闪光、天体包的星星与流星。
 *
 * 纯表现：颜色用的 `pseudorandom_element(self.colours)` 不带 seed，走的是全局 `math.random`，
 * 不碰任何 `pseudoseed` 流，所以这里用 `Math.random`。
 *
 * 挂在 `G.ROOM_ATTACH` 上（`attach`），以房间为中心；`fill` 时在整块区域里随机出生，
 * 区域是房间尺寸减去 `padding`（原作 `T.w = major.T.w - padding`，负的 padding 往外扩）。
 * 原作经 `ROOM_ATTACH` 在 `I.MOVEABLE` 里画——**在所有 UIBox 与卡之前**，所以深度紧贴背景之上。
 */
import type { GameObjects, Scene } from 'phaser';

import type { Colour } from '../ui/colours';
import { TILE_H, TILE_W, toPx } from './coords';

export type ParticlesConfig = {
    timer?: number;
    scale?: number;
    initialize?: boolean;
    lifespan?: number;
    speed?: number;
    padding?: number;
    colours: Colour[];
    fill?: boolean;
    max?: number;
    vel_variation?: number;
    /** `pulse_max`：不管 `max`，先连着出这么多颗（最多 20），`attention_text` 盖色块时的那一阵碎屑 */
    pulse_max?: number;
    /**
     * `attach`：跟着哪个东西走（`bond = 'Strong'`，以它的中心为原点）。返回它当前的矩形（tile）。
     * 不给就是挂 `G.ROOM_ATTACH`（开包的全屏粒子）
     */
    attach?: () => { x: number; y: number; w: number; h: number };
    /** 画在哪一层。缺省 −1（紧贴背景之上） */
    depth?: number;
};

type Particle = {
    dir: number;
    facing: number;
    age: number;
    velocity: number;
    r_vel: number;
    scale: number;
    colour: Colour;
    offset: { x: number; y: number };
};

const rgb = (c: Colour) => (Math.round(c[0] * 255) << 16) | (Math.round(c[1] * 255) << 8) | Math.round(c[2] * 255);

export class Particles {
    private readonly g: GameObjects.Graphics;
    private readonly particles: Particle[] = [];
    private readonly timer: number;
    private readonly lifespan: number;
    private readonly speed: number;
    private readonly scale: number;
    /** `max`：`start_materialize` 在半程把它置 0，不再出新的 */
    max: number;
    private readonly velVariation: number;
    private readonly pulseMax: number;
    private pulsed = 0;
    private readonly fill: boolean;
    private readonly w: number;
    private readonly h: number;
    private readonly attach?: () => { x: number; y: number; w: number; h: number };
    private readonly colours: Colour[];
    /** `G.TIMERS.REAL` 的对应物：场景时间（秒） */
    private now: number;
    private lastRealTime: number;
    /** `fade_alpha`：1 = 全透明。`fade(delay, to)` 线性缓动过去 */
    private fadeAlpha = 0;
    private fadeFrom = 0;
    private fadeTo = 0;
    private fadeT0 = 0;
    private fadeDur = 0;
    private removeAt = Infinity;
    private readonly onUpdate = (time: number, delta: number) => this.tick(time / 1000, delta / 1000);

    constructor(private readonly scene: Scene, config: ParticlesConfig) {
        this.timer = config.timer ?? 0.5;
        this.lifespan = config.lifespan ?? 1;
        this.speed = config.speed ?? 1;
        this.scale = config.scale ?? 1;
        this.max = config.max ?? 1e15;
        this.velVariation = config.vel_variation ?? 1;
        this.pulseMax = Math.min(20, config.pulse_max ?? 0);
        this.fill = !!config.fill;
        this.colours = config.colours;
        const padding = config.padding ?? 0;
        this.attach = config.attach;
        const major = config.attach?.() ?? { x: 0, y: 0, w: TILE_W, h: TILE_H };
        this.w = this.fill ? major.w - padding : 0;
        this.h = this.fill ? major.h - padding : 0;
        this.now = scene.time.now / 1000;
        this.lastRealTime = this.now - this.timer;
        this.g = scene.add.graphics().setDepth(config.depth ?? -1);

        // `initialize`：先空跑 60 步、每步 1/4 秒，开包那一刻屏幕上就已经铺满
        if (config.initialize) {
            for (let i = 0; i < 60; i++) {
                this.lastRealTime -= 15 / 60;
                this.update();
                this.move(15 / 60);
            }
        }
        scene.events.on('update', this.onUpdate);
    }

    /** `Particles:fade(delay, to)`：`fade_alpha` 在 `delay` 秒里缓动到 `to`（默认 1 = 淡出） */
    fade(delay: number, to = 1): void {
        this.fadeFrom = this.fadeAlpha;
        this.fadeTo = to;
        this.fadeT0 = this.now;
        this.fadeDur = delay;
    }

    /** 建好就淡入（`fade_alpha = 1` 再 `fade(1, 0)`），各个开包分支都这么写 */
    fadeIn(): this {
        this.fadeAlpha = 1;
        this.fade(1, 0);
        return this;
    }

    /** `end_consumeable`：淡出 `delay` 秒，1 秒后移除 */
    fadeOutAndRemove(delay = 1): void {
        this.fade(delay, 1);
        this.removeAt = this.now + 1;
    }

    private update(): void {
        let added = 0;
        while (this.now > this.lastRealTime + this.timer && (this.particles.length < this.max || this.pulsed < this.pulseMax) && added < 20) {
            this.lastRealTime += this.timer;
            if (this.pulsed <= this.pulseMax) this.pulsed++;
            this.particles.push({
                dir: Math.random() * 2 * Math.PI,
                facing: Math.random() * 2 * Math.PI,
                age: 0,
                velocity: this.speed * (this.velVariation * Math.random() + (1 - this.velVariation)) * 0.7,
                r_vel: 0.2 * (0.5 - Math.random()),
                scale: 0,
                colour: this.colours[Math.floor(Math.random() * this.colours.length)]!,
                offset: { x: this.fill ? (0.5 - Math.random()) * this.w : 0, y: this.fill ? (0.5 - Math.random()) * this.h : 0 },
            });
            added++;
        }
    }

    private move(dt: number): void {
        const S = this.scale;
        const L = this.lifespan;
        for (let i = this.particles.length - 1; i >= 0; i--) {
            const p = this.particles[i]!;
            p.age += dt;
            p.scale = Math.min(2 * Math.min((p.age / L) * S, S * ((L - p.age) / L)), S);
            if (p.scale < 0) {
                this.particles.splice(i, 1);
                continue;
            }
            p.offset.x += p.velocity * Math.sin(p.dir) * dt;
            p.offset.y += p.velocity * Math.cos(p.dir) * dt;
            p.facing += p.r_vel * dt;
            p.velocity = Math.max(0, p.velocity - p.velocity * 0.07 * dt);
        }
    }

    private tick(now: number, dt: number): void {
        this.now = now;
        if (this.fadeDur > 0) {
            const k = Math.min(1, (now - this.fadeT0) / this.fadeDur);
            this.fadeAlpha = this.fadeFrom + (this.fadeTo - this.fadeFrom) * k;
        }
        if (now >= this.removeAt) {
            this.destroy();
            return;
        }
        this.update();
        this.move(dt);
        this.draw();
    }

    /** 每颗是一个绕自己中心转的实心方块，边长 = 当前 `scale`（tile） */
    private draw(): void {
        const g = this.g.clear();
        const major = this.attach?.();
        const cx = major ? major.x + major.w / 2 : TILE_W / 2;
        const cy = major ? major.y + major.h / 2 : TILE_H / 2;
        for (const p of this.particles) {
            const a = p.colour[3] * (1 - this.fadeAlpha);
            if (a <= 0 || p.scale <= 0) continue;
            const half = p.scale / 2;
            const c = Math.cos(p.facing);
            const s = Math.sin(p.facing);
            const x = cx + p.offset.x;
            const y = cy + p.offset.y;
            const corner = (dx: number, dy: number) => ({ x: toPx(x + dx * c - dy * s), y: toPx(y + dx * s + dy * c) });
            g.fillStyle(rgb(p.colour), a);
            g.fillPoints([corner(-half, -half), corner(half, -half), corner(half, half), corner(-half, half)] as never, true);
        }
    }

    destroy(): void {
        this.scene.events.off('update', this.onUpdate);
        this.g.destroy();
    }
}
